apmproxy/apmserver.go

```diff
@@ -313,3 +314,64 @@ func (c *Client) WaitForFlush() <-chan struct{} {
 	defer c.flushMutex.Unlock()
 	return c.flushCh
 }
+
+func (c *Client) forwardAPMDataByType(ctx context.Context, apmData APMData, batch *BatchData) error {
+	switch apmData.Type {
+	case Agent:
+		if len(c.metadata) == 0 {
+			metadata, err := ProcessMetadata(apmData)
+			if err != nil {
+				return fmt.Errorf("failed to extract metadata from agent payload %w", err)
+			}
+			c.setMetadata(metadata)
+		}
+		return c.PostToApmServer(ctx, apmData)
+	case Lambda:
+		if err := batch.Add(apmData); err != nil {
+			c.logger.Warnf("Dropping data due to error: %v", err)
```

Comment:
This worries me. If we are going to send a transaction for OOMing/timed-out Lambda invocations, then silently dropping data like this may skew results. Can we find a way to push back on the Lambda runtime, so it backs off and retries sending the logs later?

Reply:
As per the current logic, we try to send data as soon as the batch reaches 90% usage or ages to 10 seconds (from the time the first entry was added) so the case when we drop data will occur only if APM-Server has failed for more than one tries. In addition to this, pushing back on lambda in such cases sounds reasonable. I think it should be possible to do this, is it okay to address this in a separate PR?

Reply:
Sounds reasonable - thanks.